examples/pytorch/graphsage/dist/train_dist_unsupervised.py

```diff
@@ -420,65 +251,68 @@ def run(args, device, data):
         iter_tput = []
 
         start = time.time()
-        # Loop over the dataloader to sample the computation dependency graph as a list of
-        # blocks.
-        for step, (pos_graph, neg_graph, blocks) in enumerate(dataloader):
-            tic_step = time.time()
-            sample_t.append(tic_step - start)
-
-            pos_graph = pos_graph.to(device)
-            neg_graph = neg_graph.to(device)
-            blocks = [block.to(device) for block in blocks]
-            # The nodes for input lies at the LHS side of the first block.
-            # The nodes for output lies at the RHS side of the last block.
-
-            # Load the input features as well as output labels
-            batch_inputs = blocks[0].srcdata["features"]
-            copy_time = time.time()
-            feat_copy_t.append(copy_time - tic_step)
-
-            # Compute loss and prediction
-            batch_pred = model(blocks, batch_inputs)
-            loss = loss_fcn(batch_pred, pos_graph, neg_graph)
-            forward_end = time.time()
-            optimizer.zero_grad()
-            loss.backward()
-            compute_end = time.time()
-            forward_t.append(forward_end - copy_time)
-            backward_t.append(compute_end - forward_end)
-
-            # Aggregate gradients in multiple nodes.
-            optimizer.step()
-            update_t.append(time.time() - compute_end)
-
-            pos_edges = pos_graph.number_of_edges()
-            neg_edges = neg_graph.number_of_edges()
-
-            step_t = time.time() - start
-            step_time.append(step_t)
-            iter_tput.append(pos_edges / step_t)
-            num_seeds += pos_edges
-            if step % args.log_every == 0:
-                print(
-                    "[{}] Epoch {:05d} | Step {:05d} | Loss {:.4f} | Speed (samples/sec) {:.4f} | time {:.3f} s"
-                    "| sample {:.3f} | copy {:.3f} | forward {:.3f} | backward {:.3f} | update {:.3f}".format(
-                        g.rank(),
-                        epoch,
-                        step,
-                        loss.item(),
-                        np.mean(iter_tput[3:]),
-                        np.sum(step_time[-args.log_every :]),
-                        np.sum(sample_t[-args.log_every :]),
-                        np.sum(feat_copy_t[-args.log_every :]),
-                        np.sum(forward_t[-args.log_every :]),
-                        np.sum(backward_t[-args.log_every :]),
-                        np.sum(update_t[-args.log_every :]),
+        with model.join():
```

Comment:
Why need this?

Reply:
to avoid potential hang or error during collective communication in the backend(https://pytorch.org/tutorials/advanced/generic_join.html#what-is-join), though it should be rare as we specify `force_even=True`. This could be a best practice?

Reply:
Yes, I recommend using the best practice by PyTorch.

Reply:
the reason why I define a dummy `join` in `DistSAGE` is to make the same script runnable in standalone mode.